File: src/main/java/edu/kit/iti/formal/stvs/model/expressions/TypeBool.java
package edu.kit.iti.formal.stvs.model.expressions;

import java.util.Optional;

/**
 * Runtime-representation for boolean types.
 * This is a singleton since this class does not have any state.
 *
 * @author Philipp
 */
public class TypeBool implements Type {

  public static final TypeBool BOOL = new TypeBool();

  private TypeBool() {}

  @Override
  public <R> R match(TypeIntegerHandler<R> matchIntType, TypeBooleanHandler<R> matchBoolType,
      TypeEnumHandler<R> matchEnumType) {
    return matchBoolType.handle();
  }

  @Override
  public boolean checksAgainst(Type other) {
    return other.match(() -> false, () -> true, (otherEnum) -> false);
  }

  @Override
  public String getTypeName() {
    return "BOOL";
  }

  @Override
  public Optional<Value> parseLiteral(String literal) {
    if ("true".equalsIgnoreCase(literal)) {
      return Optional.of(ValueBool.TRUE);
    }
    if ("false".equalsIgnoreCase(literal)) {
      return Optional.of(ValueBool.FALSE);
    }
    return Optional.empty();
  }

  @Override
  public Value generateDefaultValue() {
    return ValueBool.FALSE;
  }

  public String toString() {
    return "TypeBool";
  }

  @Override
  public boolean equals(Object obj) {
    return obj instanceof TypeBool;
  }
}

File: stvs/src/main/kotlin/edu/kit/iti/formal/stvs/model/expressions/TypeBool.kt
/* *****************************************************************
 * This file belongs to verifaps-lib (https://verifaps.github.io).
 * SPDX-License-Header: GPL-3.0-or-later
 * 
 * This program isType free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program isType distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a clone of the GNU General Public
 * License along with this program.  If not, see
 * <http://www.gnu.org/licenses/gpl-3.0.html>.
 * *****************************************************************/
package edu.kit.iti.formal.stvs.model.expressions

/**
 * Runtime-representation for boolean types.
 * This is a singleton since this class does not have any state.
 *
 * @author Philipp
 */
object TypeBool : Type {
    override fun checksAgainst(other: Type): Boolean =
        other.match({ false }, { true }, { otherEnum: TypeEnum? -> false })!!

    override val typeName: String
        get() = "BOOL"

    override fun parseLiteral(literal: String): Value? {
        if ("true".equals(literal, ignoreCase = true)) {
            return ValueBool.TRUE
        }
        if ("false".equals(literal, ignoreCase = true)) {
            return ValueBool.FALSE
        }
        return null
    }

    override fun generateDefaultValue(): Value = ValueBool.FALSE

    override fun toString(): String = "TypeBool"
}